File: src/unpushed.py
"""End-of-run unpushed/dirty work scanner.

Per design v2: just before normal termination, scan ``$GH_README_REPOS_DIR/*/``.
For each clone, run ``git status --porcelain`` and ``git rev-list @{u}..HEAD``
(skip clones with no upstream). If any clone has a dirty working tree OR
unpushed commits, return the list of offending paths. ``main()`` uses this to
decide between exit 0 and exit 2.
"""
from __future__ import annotations

import subprocess
from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class UnpushedFinding:
    """One repo with unpushed work."""

    path: Path
    dirty: bool
    unpushed_commits: int  # 0 if upstream missing or all pushed


def _is_git_repo(path: Path) -> bool:
    return (path / ".git").exists()


def _has_upstream(repo_dir: Path) -> bool:
    result = subprocess.run(
        ["git", "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}"],
        cwd=repo_dir,
        check=False,
        capture_output=True,
        text=True,
    )
    return result.returncode == 0


def _is_dirty(repo_dir: Path) -> bool:
    result = subprocess.run(
        ["git", "status", "--porcelain"],
        cwd=repo_dir,
        check=False,
        capture_output=True,
        text=True,
    )
    return bool(result.stdout.strip())


def _unpushed_count(repo_dir: Path) -> int:
    result = subprocess.run(
        ["git", "rev-list", "--count", "@{u}..HEAD"],
        cwd=repo_dir,
        check=False,
        capture_output=True,
        text=True,
    )
    if result.returncode != 0:
        return 0
    try:
        return int(result.stdout.strip() or "0")
    except ValueError:
        return 0


def scan_repos(repos_dir: Path) -> list[UnpushedFinding]:
    """Return per-repo findings of dirty trees or unpushed commits.

    Clones without an upstream are checked only for dirtiness; their
    unpushed_commits is 0. Non-git directories are silently skipped.
    """
    if not repos_dir.exists():
        return []

    findings: list[UnpushedFinding] = []
    for child in sorted(repos_dir.iterdir()):
        if not child.is_dir() or not _is_git_repo(child):
            continue

        dirty = _is_dirty(child)
        unpushed = 0
        if _has_upstream(child):
            unpushed = _unpushed_count(child)

        if dirty or unpushed > 0:
            findings.append(
                UnpushedFinding(path=child, dirty=dirty, unpushed_commits=unpushed)
            )
    return findings

File: go/internal/unpushed/unpushed.go
// Package unpushed scans the repos cache for dirty trees / unpushed commits.
package unpushed

import (
	"context"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
)

// Finding describes one repo with unpushed work.
type Finding struct {
	Path            string
	Dirty           bool
	UnpushedCommits int
}

// Scan returns sorted findings under reposDir.
// Non-git directories silently skipped. Clones with no upstream are
// checked for dirtiness only.
func Scan(ctx context.Context, reposDir string) ([]Finding, error) {
	entries, err := os.ReadDir(reposDir)
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	var out []Finding
	for _, e := range entries {
		if !e.IsDir() {
			continue
		}
		dir := filepath.Join(reposDir, e.Name())
		if _, err := os.Stat(filepath.Join(dir, ".git")); err != nil {
			continue
		}
		dirty := isDirty(ctx, dir)
		unpushed := 0
		if hasUpstream(ctx, dir) {
			unpushed = unpushedCount(ctx, dir)
		}
		if dirty || unpushed > 0 {
			out = append(out, Finding{Path: dir, Dirty: dirty, UnpushedCommits: unpushed})
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].Path < out[j].Path })
	return out, nil
}

func runGit(ctx context.Context, dir string, args ...string) (string, int) {
	cmd := exec.CommandContext(ctx, "git", args...)
	cmd.Dir = dir
	out, err := cmd.Output()
	code := 0
	if ee, ok := err.(*exec.ExitError); ok {
		code = ee.ExitCode()
	} else if err != nil {
		code = -1
	}
	return string(out), code
}

func isDirty(ctx context.Context, dir string) bool {
	out, _ := runGit(ctx, dir, "status", "--porcelain")
	return strings.TrimSpace(out) != ""
}

func hasUpstream(ctx context.Context, dir string) bool {
	_, code := runGit(ctx, dir, "rev-parse", "--abbrev-ref", "--symbolic-full-name", "@{u}")
	return code == 0
}

func unpushedCount(ctx context.Context, dir string) int {
	out, code := runGit(ctx, dir, "rev-list", "--count", "@{u}..HEAD")
	if code != 0 {
		return 0
	}
	n, err := strconv.Atoi(strings.TrimSpace(out))
	if err != nil {
		return 0
	}
	return n
}
